### PR TITLE
[add] Merge backend models

### PR DESCRIPTION
**Pull request by adamwhat**
- DataSource: Load training data from events
- Preparator: Train a KMeans model to cluster locations
              Aggregate demand based on time and location
              Build feature vectors with weather
- Algorithm: Implemented LinearRegressionSGD, GradientBoostTree, and
             MultinomialLogisticRegression models
- Evaluator: Eval

----------------------
Squashed commit messages
d4460ce [add] add pio engine template from detrevid/predictionio-load-forecasting
95f8009 changed Engine and Preparator to koober format. Also changed DemoData so that it produced the datetime as a Long
055b5d6 Added untested DataSource.scala
7f8a32d Moved untested DataSource.scala to pio-engine
ec55363 [add] fix build for pio 0.10
2af59df Update Algorithm.scala with koober data structures.
47c46c6 fixed error in Evaluator
8fa2c48 [add] gitignore
7b5037c fix build errors in Preparator.
a4a5239 [add] fix intellij build
7572484 [cleanup] fix gitignore
5eb158f removed the change from datetime to long
a9c16e8 [cleanup] update package information
b51ead1 Change all eventTime references to DateTime type, fix expectedResult calculation
5c1b348 [add] return feature vector in preperator
726e064 DataSource data not empty error fixed, and changed appname error in Eval.
a71e89d normalizing function for the eventTime
891e9ec Updating Algorithm and Preparator to accommodate data model change.
0bdfa01 some comments on the normalizer
3151a5e track
d722824 fix preparator bug that was causing training to fail.
0835de6 fixed the normalization method
11f699a fixed the query format. Now you should input a long for the eventTime
212e5ec Partition into test and training set based on time
44b4125 1) made it so that training set and testing set are split based on time. 2) Changed the actual result to be the correct actual result to compare with
25bce5f Adding unsupervised learning logic for clustering location coordinates in training and predicting workflow.
920a20d Storing the kmeans model as a static field instead of storing the label mapping for the training data...
f71c975 fixed input query format to take a string datetime instead of a datetime object
f5d9418 [add] importing the fake data
7bc6302 [add] data generation script
2b46593 [Bug Fix] Fixed the null pointer exception caused by the kMeansModel. Storing both ML models in one Model object
268d912 Fixed the training algorithm issue by hard coding parameters(ap.* does not work).
ecd7741 bug in eval - part1
a0b7357 bug in eval part2
fc0051e bug in eval final addition
d77a5e6 Fixed the training algorithm issue by hard coding parameters(ap.* does not work).
cd36c8f incorporating weather into the engine. Assuming that data is already imported.
c5d690e More weather stuff
c169d31 Adding weather data backfill script (calling wunderground weather API for historical data).
519d8c1 Fixing some syntax issue and update model changes in UserEvent.
4bb5c42 Merge pull request #10 from jamesward/demo-data-loader
f32eb2e Updated demo data loader to include weather data as well
22f5ecb Standardize feature data so that the scaling of features does not break ML model training.
210bf13 Transform location cluster label into one-hot-encoding vector when building the entire feature vector.
c1688b1 Tuning model parameters
bcc7235 gradientBoostedTree model
64b99fa minor parameter changes
5a2c04a add travis
5032e12 setup demand dashboard
74244e0 [add] PIO engine implementation for traffic demand prediction
550c85a multiple algorithms merging
57d2430 [add] PIO engine implementation for traffic demand prediction
f65cfb2 add some webapp stuff
ac4af06 use scala for demo data loader
5edacc3 support direct pio loading
3b5d2e6 pio sink needs to be a flow
162813b Fixed a bug during eval
f687181 minot
4dd8acd changed parameters for eval
cc7719f evaluation tuning
0bda8f2 Implemented MultinomialLogisticRegression algorithm for predicting demand
a438947 Rename Algorithm file to MultinomialLogisticRegressionAlgorithm
e2b076b Merger build errors fixed
a52baa1 fixed some bugs in training
34ca071 modified data loader to take weather
ba9cc35 Increase number of iterations in algorithm parameters.
0adbbba Update numClasses for Multinomial Logistic Regression.
c7c7cc3 Tuning configuration for LinearRegressionSGD
a9a0555 Stop printing in sample prepared data from training data.
c21b98b Updated data aggregation logic to count demand via grouping by Time+LocationClusterLabelTuple instead of just by Time.
e that empty commits are commented out

### CODE REVIEW

**Comment by adamwhat** [comment]
@EatingW rebased on jamesward/koober:master

**Comment by jamesward** [comment]
I'm seeing some overwrites of the latest Demand Dashboard stuff in here.  I think those changes might have to be manually backed out.

File: pio-engine/build.sbt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import AssemblyKeys._
+
+assemblySettings
+
+name := "predictionio-load-forecasting"
+
+libraryDependencies ++= Seq(
+  "org.apache.predictionio"    %% "apache-predictionio-core" % "0.10.0-incubating" % "provided",
+  "org.apache.spark"  %% "spark-core"   % "1.3.0"           % "provided",
+  "org.apache.spark"  %% "spark-mllib"  % "1.3.0"           % "provided"
```
Comment (adamwhat) [inline pio-engine/build.sbt]: Should we use 1.5.2 as we are using Spark 1.5.2?
[here](https://mvnrepository.com/artifact/org.apache.spark/spark-core_2.10/1.5.2.2.3.4.51-1) is the latest release for 1.5.2.x

File: pio-engine/build.sbt [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+)
```
Comment (adamwhat) [inline pio-engine/build.sbt]: Also, I got `Failed to load implementation from com.github.fommil.netlib.nativesystemblas` in training. I checked the [netlib-java](https://github.com/fommil/netlib-java) GitHub page and added the dependency but has no luck.

I checked more on Github and it seems we also need to specify it in Sbt in run/compile time with something like `sbt -J"-Dcom.github.fommil.netlib.BLAS=com.github.fommil.netlib.NativeRefBLAS"  assembly
`

**Comment by adamwhat** [comment]
@jamesward updated the PR to fix the merge error

File: pio-engine/build.sbt (hunk 1 above)
Comment (jamesward) [inline pio-engine/build.sbt]: I think this depends on what the version of Spark that the version of PIO we use, depends on.

File: pio-engine/build.sbt (hunk 2 above)
Comment (jamesward) [inline pio-engine/build.sbt]: Hmmm... I'm not sure how to set sbt JAVA_OPTS with PIO.  Maybe en env var.
